python_modules/dagster/dagster/_core/instance/ref.py

```diff
@@ -455,54 +467,93 @@ def value_for_ref_item(k, v):
+
+        return (
+            self.event_storage_data.rehydrate(as_type=EventLogStorage)
+            if self.event_storage_data
+            else None
+        )
 
     @property
-    def schedule_storage(self):
-        return self.schedule_storage_data.rehydrate() if self.schedule_storage_data else None
+    def schedule_storage(self) -> Optional["ScheduleStorage"]:
+        from dagster._core.storage.schedules.base import ScheduleStorage
+
+        return (
+            self.schedule_storage_data.rehydrate(as_type=ScheduleStorage)
+            if self.schedule_storage_data
+            else None
+        )
 
     @property
-    def compute_log_manager(self):
-        return self.compute_logs_data.rehydrate()
+    def compute_log_manager(self) -> "ComputeLogManager":
+        from dagster._core.storage.compute_log_manager import ComputeLogManager
+
+        return self.compute_logs_data.rehydrate(as_type=ComputeLogManager)
 
     @property
-    def scheduler(self):
-        return self.scheduler_data.rehydrate() if self.scheduler_data else None
+    def scheduler(self) -> Optional["Scheduler"]:
+        from dagster._core.scheduler.scheduler import Scheduler
+
+        return self.scheduler_data.rehydrate(as_type=Scheduler) if self.scheduler_data else None
 
     @property
-    def run_coordinator(self):
-        return self.run_coordinator_data.rehydrate() if self.run_coordinator_data else None
+    def run_coordinator(self) -> Optional["RunCoordinator"]:
+        from dagster._core.run_coordinator.base import RunCoordinator
+
+        return (
+            self.run_coordinator_data.rehydrate(as_type=RunCoordinator)
+            if self.run_coordinator_data
+            else None
+        )
 
     @property
-    def run_launcher(self):
-        return self.run_launcher_data.rehydrate() if self.run_launcher_data else None
+    def run_launcher(self) -> Optional["RunLauncher"]:
+        from dagster._core.launcher.base import RunLauncher
+
+        return (
+            self.run_launcher_data.rehydrate(as_type=RunLauncher)
+            if self.run_launcher_data
+            else None
+        )
 
     @property
-    def secrets_loader(self):
+    def secrets_loader(self) -> "SecretsLoader":
+        from dagster._core.secrets.env_file import EnvFileLoader
+        from dagster._core.secrets.loader import SecretsLoader
+
         # Defining a default here rather than in stored config to avoid
         # back-compat issues when loading the config on older versions where
         # EnvFileLoader was not defined
         return (
-            self.secrets_loader_data.rehydrate()
+            self.secrets_loader_data.rehydrate(as_type=SecretsLoader)
             if self.secrets_loader_data
             else ConfigurableClassData(
                 "dagster._core.secrets.env_file",
                 "EnvFileLoader",
                 yaml.dump({}),
-            ).rehydrate()
+            ).rehydrate(as_type=EnvFileLoader)
```

Comment:
Some vague fear over these runtime checks causing problems in the future for weird back compat cases since we pass the instance-ref to potentially old user code servers. Hard to imagine how exactly it would be a problem though since the target must exist in the old code and should be a child of the correct class.

Reply:
>Hard to imagine how exactly it would be a problem though since the target must exist in the old code

Yeah because of this not sure what would go wrong, but possibly not thinking everything through